Adding Spacing lesson (#5177)
GoogleChrome/web.dev: src/site/content/en/learn/css/spacing/index.md

Margin is like adding a cushion around your element.
The `margin` property is shorthand for `margin-top`,
`margin-right`, `margin-bottom` and `margin-left`.

{% Img
src="image/VbAJIREinuYvovrBzzvEyZOpw5w1/ECuEOJEGnudhXW5JEFih.svg",
alt="A diagram of the four main areas of the box model.",
width="800",
height="547" %}

The `margin` shorthand applies properties in a particular order:
top, right, bottom and left.
You can remember these with trouble: TRouBLe.

{% Img
src="image/VbAJIREinuYvovrBzzvEyZOpw5w1/F8jtvl15CsAcs8Oaid2l.svg",
alt="The word 'Trouble' running downwards with T, R, B and L
extending to Top, Right, Bottom and Left.
A box with arrows showing the directions too.",
width="800",
height="320" %}

The `margin` shorthand can also be used with one, two, or three values.
Adding a fourth value lets you set each individual side.
These are applied as follows:

- One value will be applied to all sides. (i.e. `margin: 20px`)
- Two values: the first value will be applied to the top and bottom sides,
and the second value will be applied to the left and right sides.
(i.e. `margin: 20px 40px`)
- Three values: the first value is `top`,
the second value is `left` **and** `right`,

> and the third value is `bottom`. (i.e. margin: 20px 40px 30px`)

Reviewer: ```suggestion
and the third value is `bottom`. (i.e. margin: `20px 40px 30px`)
```

If you use a percentage,
the value will be calculated based on the width of your element's containing block.

This means that if your element's containing block has a width of `250px`
and your element has a `margin` value of `20%`:
each side of your element will have a computed margin of `50px`.

{% Codepen {
  user: 'web-dot-dev',
  id: 'YzNQeyb'
} %}

You can also use a value of `auto` for margin.
For block level elements with a restricted size,
an `auto` margin will take up available space in the direction that it is applied to.
A good example is this one,
from the [flexbox module](/learn/css/flexbox), where the items push away from each other.

{% Codepen {
  user: 'web-dot-dev',
  id: 'poRELbR'
} %}

Another good example of `auto` margin is a horizontally centered wrapper which has a max width.
This sort of wrapper is often used to create a consistent center column on a website.

```css
.wrapper {
	max-width: 400px;
	margin: 0 auto;
}
```

{% Codepen {
  user: 'web-dot-dev',
  id: 'XWpgZKq'
} %}

Here, margin is removed from the top and bottom (block) sides,
and `auto` shares the space between the left and right (inline) sides.

{% Aside %}

> In the last lesson on [logical properties](/learn/css/logical-properties),

Reviewer: ```suggestion
In the previous module on [logical properties](/learn/css/logical-properties),
```

you learned that instead of specifying `margin-top`, `margin-right`, `margin-bottom` and `margin-left`,
you can use `margin-block-start`, `margin-inline-end`, `margin-block-end` and `margin-inline-start`.
{% endAside %}

### Negative margin

Negative values can also be used for margin.
Instead of adding space between adjacent sibling elements,
it will **reduce space** between them.
This can result in overlapping elements,
if you declare a negative value that's more than the available space.

{% Codepen {
  user: 'web-dot-dev',

>   id: 'YzNQeLP'

Reviewer: demo suggestion: Update the background-color on the second element to `background: hsl(20 92% 95% / 60%);` so its a bit more clear in the overlap

Reviewer: I got into the acct and made this change!

} %}

### Margin collapse

Margin collapse is a tricky concept,
but it's something you'll run into very commonly when building interfaces.
Say you have two elements: a heading and a paragraph that both have vertical margin on them:

```html
<article>
  <h1>My heading with teal margin</h1>
  <p>A paragraph of text that has blue margin on it, following the heading with margin.</p>
</article>
```

```css
h1 {
	margin-bottom: 2rem;
}

p {
	margin-top: 3rem;
}
```

{% Codepen {
  user: 'web-dot-dev',

>   id: 'OJWgvMx'

Reviewer: An idea for this demo:

Let the user change the `margin-bottom` on the h1 with a slider to directly see how margin collapse works.

Reviewer 2: Good idea: sorted

} %}

At first glance,
you would be forgiven for thinking that the paragraph will be spaced `5em` from the heading,
because `2rem` and `3rem` combined calculate to `5rem`.
Because **vertical margin collapses**, though, the space is actually `3rem`.

Margin collapse works by selecting the largest value of two adjoining elements
with vertical margin set on the adjoining sides.
The bottom of the `h1` meets the top of the `p`,
so the largest value of the `h1`'s bottom margin and the `p`'s top margin is selected.
If the `h1` were to have `3.5rem` of bottom margin,
the space between them both would then be `3.5rem` because that is larger than `3rem`.
Only block margins collapse, not inline (horizontal) margins.

> Margin collapse is rooted back to when the web was mostly just documents.

Reviewer: ```suggestion
This concept is rooted back to when the web was mostly just documents.
```

These 3 paragraphs all start with "margin collapse", so im thinking on how to change it. I also this this could be an aside element to break up the text blocks

Author: good call, I've made it an aside in my recent commit.

{% Codepen {
  user: 'web-dot-dev',
  id: 'eYgRrzz'
} %}

The `padding` property is shorthand for `padding-top`, `padding-right`, `padding-bottom` and `padding-left`.
Just like `margin`, `padding` has logical properties, too:
`padding-block-start`, `padding-inline-end`, `padding-block-end` and `padding-inline-start`.

## Positioning

Also covered  in the [layout](/learn/css/layout/) module,
if you set a value for `position` that is anything other than `static`,
you can space elements with the `top`, `right`, `bottom` and `left` properties.
There are some differences with how these directional values behave:

- An element with `position: relative` will maintain its place in the document flow,
even when you set these values.
They will be relative to your element's position too.
- An element with `position: absolute`
will base the directional values from the relative parent's position.
- An element with `position: fixed`
will base the directional values on the viewport.
- An element with `position: sticky`
will only apply the directional values when it is in its docked/stuck state.

{% Codepen {
  user: 'web-dot-dev',
  id: 'NWdNGZB'
} %}

> In the [logical properties](/learn/css/logical-properties) lesson,

Reviewer: ```suggestion
In the [logical properties](/learn/css/logical-properties) module,
```

> you learned about the `inset` property,

Reviewer: ```suggestion
you learn about the `inset` property,
```

which allows you to set directional values that honor writing mode.

The `inset` order of properties is the same as TRouBLe with `margin` and `padding`
but instead of `top`, `right`, `bottom` and `left`,
it is `block-start`, `inline-end`, `block-end` and `inline-start`

## Grid and flexbox

Lastly, in both grid and flexbox you can use the `gap` property to create space *between* child elements.
The `gap` property is shorthand for `row-gap` and `column-gap`,
it accepts one or two values, which can be lengths or percentages.
You can also use keywords such as `unset`, `initial` and `inherit`.
If you define only one value,
the same `gap` will be applied to both the rows and columns,
but if you define both values,
the first value is `row-gap` and the second value is `column-gap`.

With both flexbox and grid, you can also create space using their distribution and alignment capabilities,
which we cover in the

> [grid lesson](/learn/css/grid) and

Reviewer: ```suggestion
[grid module](/learn/css/grid) and
```

> [flexbox lesson](/learn/css/flexbox).

Reviewer: ```suggestion
[flexbox module](/learn/css/flexbox).
```

which provides an inline space.
Be careful though,
because the non-breaking aspect of this character stitches the two elements together,
which can result in odd behaviour.

{% Aside %}
Use HTML elements to add space only when the element helps with the understanding of the document.
For example, an `<hr>` doesn't just add space,
it creates a logical separation of two chunks of content.
If you just want a line with space around it,
adding a border with CSS might be more appropriate.
{% endAside %}

## Margin

If you want to add space to the outside of an element,
you use the `margin` property.
Margin is like adding a cushion around your element.
The `margin` property is shorthand for `margin-top`,
`margin-right`, `margin-bottom` and `margin-left`.

{% Img
src="image/VbAJIREinuYvovrBzzvEyZOpw5w1/PJfvU46GbvkmIYD5ueTb.png",
alt="A diagram of the four main areas of the box model.",
width="800",
height="547" %}

The `margin` shorthand applies properties in a particular order:
top, right, bottom and left.
You can remember these with trouble: TRouBLe.

{% Img
src="image/VbAJIREinuYvovrBzzvEyZOpw5w1/oBvH9AMffeuMAy9XYLYF.png",

> alt="The word "Trouble" running downwards with T, R, B and L

Reviewer 3: ```suggestion
alt="The word 'Trouble' running downwards with T, R, B and L
```
